net/mlx5: separate mlx5_rx_burst into a regular vs optimized no_sges versions (ETH and IPoIB)

tomerfiliba opened the pull request:
Backport mlx5_rx_replenish_bulk_mbuf from DPDK 17.11

tomer-weka committed:
net/mlx5: separate mlx5_rx_burst into a regular vs optimized no_sges versions (ETH and IPoIB)

Backport mlx5_rx_replenish_bulk_mbuf from DPDK 17.11

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
+
+/**
+ * Replenish buffers for RX in bulk.
+ * Backported from DPDK 17.11
```
no need for the backport comment

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
+	unsigned int i;
+
+	assert(rxq->refill_count >= MLX5_VPMD_RXQ_RPLNSH_THRESH);
+
```
redundent line

tomerfiliba replied:
the assert? we took it from http://dpdk.org/browse/dpdk/tree/drivers/net/mlx5/mlx5_rxtx_vec.h?h=v17.11#n109

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
+
+		if (rte_mempool_get_bulk(rxq->mp, (void *) elts, batch_size) < 0) {
+			//RTE_LOG(ERR, PMD, "PMD: RX ring couldn't be replenished, missing %d pkts (batch_size=%d)\n",
+			//        rxq->refill_count, batch_size);
```
remove the commented lines

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
+ * @param rxq
+ *   Pointer to RX queue structure.
+ * @param rel_data_offset
+ *   Should be 0 for ETH, GRH_HDR_LEN for IPoIB
```
need to explain what this value means, and not with what values it should be set.

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+	rte_prefetch0(cqe);
+	rte_prefetch0(cqe+1);
+	rte_prefetch0(cqe+2);
+	// prefetch the next cache line (no point in prefetching the current)
```
comment style is wrong.

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+
+		if ((idx & 4) == 0) {
+			struct rte_mbuf **next_mbufs_cache_line = &(*rxq->elts)[(idx + (64/sizeof(struct rte_mbuf *))) & wqe_cnt];
+			// we start a new cache line, prefetch the next one
```
comment style is wrong. there are more cases below

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+		rte_prefetch0(cqe+3); // we're allowed to prefetch invalid addresses
+		rte_prefetch0((*rxq->elts)[next_idx]); // prefetch the next mbuf header
+		seg = rep;
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+			++rxq->stats.idropped;
+			goto skip;
+		}
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+		struct rte_mbuf *pkt = seg;
+		assert(len >= (rxq->crc_present << 2));
+		/* Update packet information. */
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+		NB_SEGS(pkt) = 1;
+		PORT(pkt) = port;
+		NEXT(pkt) = NULL;
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+		}
+		if (rxq->crc_present)
+			len -= ETHER_CRC_LEN;
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+		// sges_n == 0 so these two are equal
+		PKT_LEN(pkt) = len;
+		DATA_LEN(pkt) = len;
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+	}
+	if (unlikely((i == 0) && (rq_ci == rxq->rq_ci)))
+		return 0;
+
```
redundant blank line

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+	rxq->refill_count += i;
+	if (rxq->refill_count >= MLX5_VPMD_RXQ_RPLNSH_THRESH) {
+        // RTE_LOG(ERR, PMD, "PMD: will refill %d pkts starting at %d. first_idx=%d i=%d rq_ci=%d wqe_cnt=%d\n",
+        //         rxq->refill_count, rxq->rq_ci, first_idx, i, rq_ci, wqe_cnt);
```
remove commented lines

shahafsh commented on drivers/net/mlx5/mlx5_rxtx.c:
```diff
@@ -2538,6 +2723,117 @@ mlx5_rx_burst_ipoib(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
+uint16_t
+mlx5_rx_burst_ipoib_no_sges(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n) {
+	assert(rxq->sges_n == 0);
+
```
same comment as in eth burst function.

redundant extra lines.
wrong comment style.

shahafsh reviewed (commented):
Hi Tomer,

i added few comment regarding to coding style.
even though this is not upstream code i would like to keep on high standard, for us to be able to maintain it easily.

my big question is -
i understand the big improvement by replenish set of buffers in bulk.

however how much gain you see when not using sge in compare to the version which does?

if it is not large, i suggest to modify the current eth and ipoib burst to replenish buffers in bulk and not introduce new set of data path functions.

shahafsh commented:
@tomerfiliba 

please see some comments and question

tomerfiliba commented:
@shahafsh, the replenish logic only works if `sges_n == 0`. this is also the case in the vectorized version we took it from

tomer-weka committed:
net/mlx5: CR fixes for the optimized no_sges version

tomerfiliba commented:
@shahafsh i removed most of the blank lines and fixed the comments, however, some of the blank lines are there for readability (marking the boundaries of code blocks)

tomer-weka committed:
net/mlx5: move assert to after the variable declaration

shahafsh commented:
@tomerfiliba 

squashed the CR fix and push to master branch with some commit log rewroding.

thanks.
